### ISSUE TITLE
python 3 support/default

### ISSUE THREAD

**Issue by PhillipSz** [dagwieers/dstat#118]
##### SUMMARY

Please move to python 3 as will die in 2020.

##### ISSUE TYPE

 - Feature Idea

**Comment by rpodgorny** [comment]
+1

see my branch at https://github.com/rpodgorny/dstat/tree/python3 for some preliminary work. the base dstat works, now it's time to convert plugins. ...but the problem so far is the import logic - it's one big mess. :-(

**Comment by dagwieers** [comment]
Indeed, I wouldn't port over what we have.

Instead we need to refactor the plugin mechanism, and preferably the main loop to accomodate both python2 _and_ python3 (with a focus on python3).

**Comment by dagwieers** [comment]
BTW Dstat was my first Python program, and that probably shows. The core has been largely a one-man project and did get little review from experienced python-programmers (some of the internal details are very weird).

So a redesign would also be very useful to clean up a lot of the logic that is a result of inexperience with Python. And it would also help in making plugins easier to write.

**Comment by rpodgorny** [comment]
well, in my opinion, porting and refactoring existing code is usually a better solution as the complete rewrites are often overthinked and/or left unfinished.

i think the best first step would be to drop support for ancient python versions (<2.7?). from there, one can start to improve the code step by step.

**Comment by PhillipSz** [comment]
I would also think about going python 3 only. Everyone, who will install dstat now on a recent distro has python 3. Those who not, can still use the legacy code. 3 only will bing much cleaner code.

**Comment by dagwieers** [comment]
@PhillipSz @rpodgorny I like the fact that I can drop dstat.py on almost any Linux system and can run it. If you go python3-only a lot of (legacy) devices fall out of that category.

BTW Dstat supported python 1.5 until a few years ago (until RHEL2.1 went out of extended support).

**Comment by PhillipSz** [comment]
well okey. But then we have to keep the python2 shebang because almost on all systems python == python2 (execpt Arch). So we will almost never have python3 as default or what do you think? And making two files is also not the way to go, right?

**Comment by rpodgorny** [comment]
not really, if you decide to drop really old pythons, the __future__ import will, hopefully, take care of the rest. at least for python2.7, imho.

**Comment by eshizhan** [comment]
My little job at https://github.com/eshizhan/dstat. I'm not testing all plugins, but some of commons should be fine with python 2.6+ and 3.2+.🌹

**Comment by jubalh** [comment]
So what exactly is the current state of this?

Also thought about rewriting this in python3.

It seems @rpodgorny and @eshizhan have both started to port to pythong3/make it work with both python2 and python3.
Seems like bundling that effort would be a better thing to do.

**Comment by praiskup** [comment]
@rpodgorny / @eshizhan could you please submit PR if you moved far enough?  Or what's the status?

**Comment by methane** [comment]
I think splitting repository or creating py3 branch is better than 2/3 compatible way.

* 2/3 compatible code is hard to write and maintain.
* Development speed of master (py2) branch is slow.  It's easy to port changes in master branch to py3 branch.
* There are only less than two years until Python 2 EOL
* Some plugins will be require dependency differ than Python 2.

**Comment by praiskup** [comment]
> I think splitting repository or creating py3 branch is better than 2/3 compatible way.

It is usually OK to support at least python 2.7 and python 3.  Is anything < python 2.6 needed?

> 2/3 compatible code is hard to write and maintain
> Some plugins will be require dependency differ than Python 2.

Can you be more concrete?

**Comment by methane** [comment]
> It is usually OK to support at least python 2.7 and python 3. Is anything < python 2.6 needed?

I meant dropping 2.7 support in py3 branch.  Master branch keep 2.7 support (until 2020).

> Can you be more concrete?

It's hard to be concrete because there are massive small parts.
But keep 2.7 support means it can't use any new Python 3 features and stdlibs.

Additionally, it's difficult to test all plugins on Travis.
If we keep current single branch, we may break some plugins while rewriting.

Keeping "dstat for Python 2" as current state and creating new branch (or repo) for
"dstat for Python 3" may makes it easy to port more quickly.

**Comment by praiskup** [comment]
> keep 2.7 support means it can't use any new Python 3 features and stdlibs.

This doesn't help to understand; why the python3 features/libs are needed?  Can those features be named here so we have a reasonable argument?

> Additionally, it's difficult to test all plugins on Travis.

Please specify why.  I can imagine there are issues, but referencing/specifying the issues concretely should help the potential PR submitter to better understand what's the real showstopper.

--
If anything, more logical way would seem to be flip master against Python3, and keep compat branch for Python2.7/Python2.6 (and backpatch when needed). But FWIW, I would be -1 if there's unclear justification.

**Comment by methane** [comment]
> This doesn't help to understand; why the python3 features/libs are needed? Can those features be named here so we have a reasonable argument?

You can check Django codebase; they drop Python 2 support and simplify code.
Django 2.0 runs faster than Django 1.11 too.
Or you can see https://pythonhosted.org/six/ for more examples.


>> Additionally, it's difficult to test all plugins on Travis.
>
> Please specify why. I can imagine there are issues, but referencing/specifying the issues concretely should help the potential PR submitter to better understand what's the real showstopper.

Because they arn't tested on Travis-CI for now.
There are many types of plugins.  Some of them requires external servers.

---

> If anything, more logical way would seem to be flip master against Python3, and keep compat branch for Python2.7/Python2.6 (and backpatch when needed).

I don't care which should be master for now.  Anyway, Python 3 should be master at last.

> But FWIW, I would be -1 if there's unclear justification.

Is there any clear justification about keeping 2/3 compat in single branch, with risk of breaking existing code?

In many projects, keeping 2/3 compat is best practice because maintaining two branches has more cost than writing 2/3 code.
But in case of this project, master branch is very stable.  There are only few commits on master branch a year.

My main point is "easy to contribute".  Many contributors may use only Python 3.

* We can use tools like 2to3 without risk of breaking existing Python 2 code.  It produce very large diff.
* We can merge it without test / precise review, because it didn't run on Python 3 at first.  It don't make something worse.
* After it, we start fixing broken parts;  In this stage, we can ignore about Python 2, and focus only Python 3.

**Comment by kapsh** [comment]
Hello, I saw your discussion about 2-3 compatible Python code here. 

@praiskup I can point to some of changes that will make support a nightmare. For starters.

1. Incompatible `dict` methods. You can't just use `items` or `keys`, you should keep track on usage: am I iterating here? Then it's maybe fine. I want to save this? Oh, just add list() here. And at some point code eats 2 GiB of RAM on py2, because `iteritems` should have been used.

2. Different layout of stdlib. Wanna use `configparser` or `httplib`? Or save memory by using `imap` and `izip`? Say hello to ugly try-new-name-except-ImportError-import-newname everywhere. 

3. Strings. You wrapped all in `unicode` and decoded/encoded all input and it even worked (most of the time, some libraries live in ascii world anyway). But it won't on py3 ofc.

4. Don't you ever forget to put `from __future__ import print, division, etc.` in each file.

5. AFAIK metaclass usage syntax and iterator protocol for user classes changed and there is no way to use them without 3rd-party libs like `six`. 

6. *(sobs)* enough.

And (briefly) about convenient things from py3 which cannot be used in 2:

- yield from,
- simple super(),
- unpacking to *var (can't remember how it is called),
- f-strings,
- pathlib,
- lru_cache,
- statistics,
- enum,
- etc.

**Comment by praiskup** [comment]
Regarding Python 2, vs 3 compat-nightmare - it really sounds like FUD to me, but I agree that one would have to have CI in place to be sure that no accidental mistake happens.  For testing, there's unittests' mock thing...  which can be used to create the abstraction/emulation for stuff which can not be tested in travis environment.  Just saying that I'm not convinced anyone should be blocked to write a compat PR.

**Comment by methane** [comment]
> Regarding Python 2, vs 3 compat-nightmare - it really sounds like FUD to me,

I maintain some projects with py2/3 compatibility.  It's best practice for several years.  It was better than "two branch" model for most projects.  I don't want FUD for it, while I think everyone should adomit keeping py 2/3 compat requires *significant* effort.

I only talk about in case of this project for now.

* "Maintain two branch" is easy for this project, because "dstat for Python 2" is very stable.  And there are only less than two years until PY2 EOL.
* And since there are various plugins, "avoid breaking existing code" is much harder than "maintain two branch".

>but I agree that one would have to have CI in place to be sure that no accidental mistake happens.

Do you mean blocking any effort until full tests?
My recommended roadmap is:

1. Make new branch, (devel or py3)
1. Do 2to3 (or futurize if we keep py2 compatibility in the branch)
1. Add test and fix bugs...

If we keep py2 compatibility, when should we recommend people to use new branch?
How is it earlier than PY2 EOL?
I don't think efforts for keeping py2 compatibility in the new branch is totally worthless.

> For testing, there's unittests' mock thing... which can be used to create the abstraction/emulation for stuff which can not be tested in travis environment. 

And unittest.mock is added in Python 3.3.  😉

**Comment by praiskup** [comment]
> Do you mean blocking any effort until setup full tests?

No! :-) CI is only needed ".. to be sure ..", for convenience of maintainers.

> And unittest.mock is added in Python 3.3.

You can use http://www.voidspace.org.uk/python/mock/

**Comment by methane** [comment]
> No! :-) CI is only needed ".. to be sure ..", for convenience of maintainers.

If we go with "single branch with py2/3 compatibility" model, how do you guarantee no breakage on Python 2?
For example, see @eshizhan 's work.
https://github.com/eshizhan/dstat/commit/f703b21fc37d1559bf419e2c5d5ac73f71990a18
"Showing  with 658 additions and 666 deletions."

More than 80 plugins are changed without CI.

> You can use http://www.voidspace.org.uk/python/mock/

Yes, but requiring third party is one of "significant effort" for keeping PY2 compatibility.

---

I still can't understand your point.  You want "one branch model" or "two branch model"?
If "one branch model", how do you guarantee no breakage on PY2?
If "two branch model", what's the benefit of keeping PY2 support?  There is PY2 stabe branch and  PY2 will be EOL on 2020.

**Comment by praiskup** [comment]
On Friday, March 23, 2018 9:40:28 AM CET INADA Naoki wrote:
> > Regarding Python 2, vs 3 compat-nightmare - it really sounds like FUD to me,
> I maintain some projects with py2/3 compatibility.  It's best practice for
> several years.  It was better than "two branch" model for most projects.  I
> don't want FUD for it, while I think everyone should adomit keeping py 2/3
> compat requires *significant* effort.
>
> I only talk about in case of this project for now.

Don't take me wrong, I just wanted to hear why **this project** should
have problems with porting.  Concrete examples in code, e.g.

> If we go with "single branch with py2/3 compatibility" model, how do you
> guarantee no breakage on Python 2?

This sounds like philosophic question;  because nothing really guarantees
you anything.  Any fix, regardless what approach is chosen, is risky.
CI only lowers the chances, as reviews do, etc.

> For example, see @eshizhan 's work.
> https://github.com/eshizhan/dstat/commit/f703b21fc37d1559bf419e2c5d5ac73f71990a18
> "Showing  with 658 additions and 666 deletions."
> 
> More than 80 plugins are changed without CI.

If there's no CI, at least try to release an alpha version and let the
users try the software.  The point is to move forward till Python 3 only
distributions (if there are any) will drop dstat completely.

> > You can use http://www.voidspace.org.uk/python/mock/
> 
> Yes, but requiring third party is one of "significant effort" for
> keeping PY2 compatibility.

Putting one 'pip' install command into travis.yml doesn't sound to be that
significant issue, to me :-).  But again, I never wanted to say that it's
easy;  I want to see concrete show-stoppers **here**.

> I still can't understand your point.  You want "one branch model" or "two
> branch model"?

- my point is to push the Python 3 work forward
- I wanted unhide concrete problems with this or that approach, for any
  patch submitter

> If "one branch model", how do you guarantee no breakage on PY2?  If "two
> branch model", what's the benefit of keeping PY2 support?

Benefits are single-point-of-failure, so single place where to fix.

> There is PY2 stabe branch and  PY2 will be EOL on 2020.

IMO it won't be easy to simply bury Python 2 in the next 10 years
(even though everyone would wish that), e.g. CentOS 7 will be here for
quite some time (Python 2.7).  Even CentOS 6 is not yet completely dead
(and there's Python 2.6), not speaking about RHEL/CentOS 5.  I interpret
this like -- if Python 2 is still a thing from project's POV -- doing the
2/3 compat work is usually the cheapest way.

**Comment by methane** [comment]
> IMO it won't be easy to simply bury Python 2 in the next 10 years
> (even though everyone would wish that), e.g. CentOS 7 will be here for
> quite some time (Python 2.7).  Even CentOS 6 is not yet completely dead
> (and there's Python 2.6), not speaking about RHEL/CentOS 5.  I interpret
> this like -- if Python 2 is still a thing from project's POV -- doing the
> 2/3 compat work is usually the cheapest way.

You can `yum install dstat` or `apt install dstat`.
Or you can download dstat.py from PY2 branch on such old machine.

Keep "old stable PY2 branch" is better to keep support old Linux, while going forward for PY3 / new Linux on new PY3 branch.

**Comment by methane** [comment]
> This sounds like philosophic question;  because nothing really guarantees
> you anything.  Any fix, regardless what approach is chosen, is risky.
> CI only lowers the chances, as reviews do, etc.

No.  If we keep PY2 branch as-is, we **can** guarantee there are no regression on PY2 branch,
because we commit such large changes only on PY3 branch.

**Comment by praiskup** [comment]
> Keep "old stable PY2 branch" is better to keep support old Linux, while going forward for PY3 / new Linux on new PY3 branch.

I just wanted to know arguments why it is so (from **this** project maintenance POV).  So you haven't convinced me, yet, but that's not necessary (stepping down from OT, poking at pcp).

**Comment by methane** [comment]
> I just wanted to know arguments why it is so (from this project maintenance POV).

Because contributors who want PY3 support may not want to setup legacy Linux to test.

**Comment by methane** [comment]
@dagwieers would you create new branch or repository?

Ubuntu is working on dropping PY2 from main apt repository.
I'm heavy user of dstat and I want PY3 version of dstat on Ubuntu 18.10.

**Comment by jubalh** [comment]
> Ubuntu is working on dropping PY2 from main apt repository.

Nothing special about Ubuntu here. Other distributions already did the same change, and the ones I checked use https://github.com/eshizhan/dstat as the new source for now. Until this is settled here.

**Comment by praiskup** [comment]
On Friday, March 23, 2018 11:13:46 AM CET INADA Naoki wrote:
> > I just wanted to know arguments why it is so (from this project maintenance
> > POV).
> 
> Because contributors who want PY3 support may not want to setup legacy Linux
> to test.

Some contributors "may not" ... either that's too general statement (totally
irrelevant to my question strictly related to this project), or you mean
yourself here...  So I guess then any step forward would be awesome, whatever
way you prefer!

**Comment by methane** [comment]
> Some contributors "may not" ... either that's too general statement (totally
> irrelevant to my question strictly related to this project),

When there are two contributors and one doesn't want to setup legacy Linux for test, "easy to contribute" is weaken by supporting Legacy Linux.
So "some contributors may not" doesn't deny my opinion.

Anyway, "setup legacy Linux and test on it" is **cost** even if the contributor doesn't dislike.

> or you mean yourself here...

Of course!  I don't want to do manual test.
"PY3 only branch" allow me to skip manual test, because there must be no degrade.

> So I guess then any step forward would be awesome, whatever way you prefer!

I prefer start unstable branch and focus on PY3 and recent Linux on the branch.

**Comment by praiskup** [comment]
On Monday, March 26, 2018 12:16:06 PM CEST INADA Naoki wrote:
> > or you mean yourself here...
> 
> Of course!  I don't want to do manual test.  "PY3 only branch" allow me to
> skip manual test, because there must be no degrade.

It's still unclear, are you going to port dstat to python3 and do the testing,
right?

**Comment by methane** [comment]
> It's still unclear, are you going to port dstat to python3 and do the testing,
> right?

It depends.  Please notice that current code is not automated tested.

For example, when I (or other contributor) run 2to3, I want merge it without testing all plugins.

Another example, I want to test some plugin manually only on one Ubuntu setup.
I may merge it without testing on Arch, Gentoo, CentOS 5,6,7, etc....
There may be compatibility problem.  But it can be fixed later.
Since PY3 "doesn't work" for now, it isn't degrade.

Of course, "All code is tested on All platform automatically" is idiomatic.
But I don't wait porting until we have it.

I don't have skill and energy to setup testing platform for (all plugins * all options * all systems).

**Comment by methane** [comment]
I'm working on https://github.com/methane/dstat/commits/py3

I removed some old (for Linux 2.4) plugins.
I confirmed some plugins work.

**Comment by methane** [comment]
> Nothing special about Ubuntu here. Other distributions already did the same change, and the ones I checked use https://github.com/eshizhan/dstat as the new source for now. Until this is settled here.

Wow, which distribution did it?

**Comment by dagwieers** [comment]
It took me less than an hour to make the codebase work on both Python 2.7 and Python 3.6.
The current master branch should work fine on Python 3. Feedback welcome.

**Comment by scottchiefbaker** [comment]
@dagwieers That's awesome. Good work sir!

**Comment by dagwieers** [comment]
@scottchiefbaker Well, I just noticed @methane did almost the exact same changes. It would have been easier if that was a PR instead :-/

**Comment by methane** [comment]
@dagwieers As I wrote above, my plan was two branch model; one for backward compatibility,
and one for Python 3.
For example, my branch drop some plugins I think it's useless for modern systems with Python 3.
And I don't use six, just support only Python 3 for new branch.

**Comment by bmurphy1976** [comment]
We've migrated every tool we use to Python3, dstat is the only one left. What's the popen3 issue?  I might be able to help but I don't know where to start.

**Comment by praiskup** [comment]
There were some undefined problems to port this to both python2/python3, but now it should be fixed according to comments above in master branch.  FWIW, Fedora seems to be moving to pcp-dstat as the alternative implementation because of the python3 compat issues..

**Comment by dagwieers** [comment]
@praiskup No, they moved to pcp-dstat because the author has its own agenda. If the issue was that this project was not compatible with Python 3, it took me 1 hour to make it work on Python 3.

They did not even contact us.

**Comment by natoscott** [comment]
@praiskup current state of dstat is covered here: https://github.com/dagwieers/dstat/issues/156

Yes, the old dstat has been replaced with a reimplementation in Fedora and RHEL due to lack of maintenance on the original.  FWLIW, the new dstat has many new features beyond python3, see:

https://www.redhat.com/en/blog/implementing-dstat-performance-co-pilot

**Comment by dagwieers** [comment]
@natoscott And it comes with quite a bagage. It's not just a python script you can drop on your Raspberry Pi or WRT router and run. If you make another product, give it another name IMO.

**Comment by dagwieers** [comment]
Alright, let's bite the bullet.

Since I have zero interest to fight a multi-billion company who likes replacing commands with their own. And I have been in the same situation before and it didn't bring any joy. And since nobody is picking up the towel, let's end this project here.

So long, and thanks for all the fish.

**Comment by scottchiefbaker** [comment]
:( 

I use the original `dstat` every day. It's clear to me that the "new" PCP based dstat is not a good replacement for my use cases. Just because RedHat chose (wrongly) to replace `dstat` doesn't mean the tool is useless. It's still used (correctly) on a lot of other distros. I would really hate to see the project get abandoned because of bad actors at RedHat.
